### PR TITLE
Add resolve context to overrule the context at resolution time

### PR DESCRIPTION
**Pull request by doozMen**
For my project it is very important that there is one global context that all registrations use and a resolve context that I can use to force a resolution to take a factory context into account. I hope we can at least discuss the option of adding this feature. Maybe the implementation is not as you would like it to be in Factory but could you point me into a direction or should I remain on a fork?

I tried to prepare for this merge request by first refactoring the way a context was used to look for the appropriate factory in #105 but that request was closed and I was pointed to develop. I removed the need to refactor the code but just add the feature.

The added unit tests` FactoryContextTests.swift:294  testResolveInContext()` should clarify how it can be used.

> Question: I also notice that develop does not seam to be up to date with main. So unclear if I should point this to main or develop?

### CODE REVIEW

**Comment by hmlongco** [comment]
Is this for testing? (e.g. force a context)? If so it might simply be easier to make the FactoryContext static variables public. If you look at the unit tests for contexts you'll see how I manipulate the context during testing to exercise all of the paths.
```swift
    func testWithArgumenst() {
        FactoryContext.isPreview = false
        FactoryContext.isTest = false
        FactoryContext.arguments = ["ARG2"]
        let service1 = Container.shared.externalContextService()
        XCTAssertEqual(service1.name, "REGISTERED")
        let service2 = Container.shared.internalContextService()
        XCTAssertEqual(service2.name, "DEBUG")
        let service3 = Container.shared.argsContextService()
        XCTAssertEqual(service3.name, "ARG")
    }
```

File: Sources/Factory/Factory/Registrations.swift [hunk 1]
```diff
@@ -309,32 +309,44 @@ internal struct FactoryOptions {
...
             }
         }
+
         if let contexts = contexts, !contexts.isEmpty {
+            let isPreview: Bool
+            let isTest: Bool
+            let isSimulator: Bool
+
+            if let resolveContext {
+                isPreview = resolveContext.isPreview || FactoryContext.isPreview
```
Comment (hmlongco) [inline Sources/Factory/Factory/Registrations.swift:329]: Not sure your "override" is doing what you think. If you're trying to force the context to be "simulator" but the FactoryContext.isTest result is true that condition will resolve before your forced comparison.

**Comment by hmlongco** [comment]
Also think this may be be what Runtime Arguments was designed to handle.

```swift
FactoryContext.setArg("light", forKey: "theme")
FactoryContext.removeArg(forKey: "theme")

myStyleSystem { StandardTheme() }
    .onArg("light") { LightTheme() }
    .onArg("dark") { DarkTheme() }
```

**Comment by hmlongco** [comment]
Download the current develop branch and check out FactoryContext.current. This should allow global overrides for any of the context states.

File: Sources/Factory/Factory/Registrations.swift (hunk 1 above)
Comment (doozMen) [reply Sources/Factory/Factory/Registrations.swift:329]: I thanks might be something I have to fix as I first made it with the inclusion of the refactor #105 but that proved to be difficult to merge with develop. So I restarted with a more modest approach. Will look into this today.

**Comment by doozMen** [comment]
Yes I tried exposing that but as it is used in a multithreaded way I could not use it to overwrite temporarily. Mainly putting back the context that it was before was problematic. The fact that it is global makes it hard to reason with and difficult as you would need to

1. store the current arguments
2. put the current arguments back

But what if inbetween 1 and 2 the current arguments change, on another thread. Then you would be putting arguments back in 2 that are not the most recent.

So instead of doing that I thought as I only need it when I do, what I call manual, resolution, not when jusing @Injectec(..) or `Container.shared.foo()` this might be cleaner as I could use in code where I need it `Container.shared.foo.resolve(in: ...) ` which does not care about the global arguments, but returns the value for the context I enforce.

So in short you are very correct and I tried to use runtimeArguments, then name indicates they are to be used for that use case. But I had trouble in a multi threaded world, which all of my projects are :).

File: Sources/Factory/Factory/Registrations.swift (hunk 1 above)
Comment (doozMen) [reply Sources/Factory/Factory/Registrations.swift:329]: A solution would be to only allow to overwrite arguments and not the isTest or other build context related things.

I will ponder a bit on that and try to propose something ...

**Comment by doozMen** [comment]
I would like to keep using @injected an as I understand it the ParameterFactory cannot be used, for good reason, with that? 

However that is not my main issue. The registrations in my project come from different libraries. Depending in which lib is added as a dependency there are more or less .onArg... registrations. Making for me the use of this context feature ideal as it offers me a way to define a dynamic context without writing that myself. I do however need to compose some resolutions of both light and dark themes as you noticed. So I got this working without the use of a parameterfactory which I found an advantage. Hence I do not think, however I love the suggestion, that using a parameter factory would make it possible accomplish my goal. 

So what do you need to see changed in order for this to be approvable?

**Comment by hmlongco** [comment]
If you want to use @Injected then all I can think of for now is....
```swift
extension Container: AutoRegistering {
    var dynamicTheme: Factory<Theme> {
        self {  self.lightTheme() }
            .onArg("light") {  self.lightTheme() }
            .onArg("dark") { self.darkTheme()  }
    }
    var lightTheme: Factory<Theme> {
        self { LightTheme()  }
    }
    var darkTheme: Factory<Theme> {
        self { DarkTheme()  }
    }
}
```
Places that want dynamicTheme inject it. Places that need to force light or dark mode inject those directly.

You can't use parameter factory with injected as there's no way to pass dynamic parameters.

**Comment by hmlongco** [comment]
If no more comments on this I'll close it?
